### PR TITLE
Delete CRDs & check finalizers in operator delete

### PR DESCRIPTION
**Pull request by arnongilboa**
Signed-off-by: Arnon Gilboa <agilboa@redhat.com>

**What this PR does / why we need it**:
`reconcileDeleteControllerDeployment` deletes the CRDs of `DataImportCron` and `DataVolume`, and returns error if any DIC/DV has a finalizer. It will get retried and controller will eventually remove the finalizers.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/operator/controller/callbacks.go [hunk 1]
```diff
@@ -79,6 +79,36 @@ func reconcileDeleteControllerDeployment(args *callbacks.ReconcileCallbackArgs)
...
 	}
 
+	args.Logger.Info("Deleting CRDs and verifing no finalizers")
+
+	crd := &extv1.CustomResourceDefinition{ObjectMeta: metav1.ObjectMeta{Name: "dataimportcrons.cdi.kubevirt.io"}}
+	if err := args.Client.Delete(context.TODO(), crd, &client.DeleteOptions{}); err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	dics := &cdiv1.DataImportCronList{}
+	if err := args.Client.List(context.TODO(), dics, &client.ListOptions{}); err != nil && !errors.IsNotFound(err) {
```
Comment (mhenriks) [inline pkg/operator/controller/callbacks.go:89]: This should be `IsNoMatchError` not `IsNotFound`

**Comment by mhenriks** [comment]
Should be "destructive" test for this that uninstalls CDI

File: pkg/operator/controller/callbacks.go [hunk 2]
```diff
@@ -79,6 +79,36 @@ func reconcileDeleteControllerDeployment(args *callbacks.ReconcileCallbackArgs)
...
+			return fmt.Errorf("DataImportCron %s has Finalizers %v", dic.Name, dic.Finalizers)
+		}
+	}
+
+	crd = &extv1.CustomResourceDefinition{ObjectMeta: metav1.ObjectMeta{Name: "datavolumes.cdi.kubevirt.io"}}
+	if err := args.Client.Delete(context.TODO(), crd, &client.DeleteOptions{}); err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	dvs := &cdiv1.DataVolumeList{}
+	if err := args.Client.List(context.TODO(), dvs, &client.ListOptions{}); err != nil && !errors.IsNotFound(err) {
```
Comment (awels) [inline pkg/operator/controller/callbacks.go:103]: Same here, check for IsNoMatchError and NotFound

**Review by awels — COMMENTED** [review]
Also agree with Michael need a destructive test here where some DICs and DVs with finalizers on them.

File: pkg/operator/controller/callbacks.go (hunk 1 above)
Comment (mhenriks) [reply pkg/operator/controller/callbacks.go:89]: I don't think list returns notfound.  Just returns an empty list in that case

File: pkg/operator/controller/callbacks.go [hunk 3]
```diff
@@ -79,6 +79,36 @@ func reconcileDeleteControllerDeployment(args *callbacks.ReconcileCallbackArgs)
 		return nil
 	}
 
+	args.Logger.Info("Deleting CRDs and verifing no finalizers")
+
+	crd := &extv1.CustomResourceDefinition{ObjectMeta: metav1.ObjectMeta{Name: "dataimportcrons.cdi.kubevirt.io"}}
+	if err := args.Client.Delete(context.TODO(), crd, &client.DeleteOptions{}); cdicontroller.IgnoreIsNoMatchError(err) != nil {
```
Comment (mhenriks) [inline pkg/operator/controller/callbacks.go:85]: NotFound is correct here

File: pkg/operator/controller/callbacks.go [hunk 4]
```diff
@@ -79,6 +79,36 @@ func reconcileDeleteControllerDeployment(args *callbacks.ReconcileCallbackArgs)
...
+		return err
+	}
+	dics := &cdiv1.DataImportCronList{}
+	if err := args.Client.List(context.TODO(), dics, &client.ListOptions{}); err != nil {
```
[the remark this answers is not in the document]
Comment (arnongilboa) [reply pkg/operator/controller/callbacks.go:89]: so in List (after we deleted the CRD) we need to accept both NoMatch and NotFound?

File: pkg/operator/controller/callbacks.go (hunk 3 above)
Comment (mhenriks) [reply pkg/operator/controller/callbacks.go:85]: Just NotFound

File: pkg/operator/controller/callbacks.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mhenriks) [reply pkg/operator/controller/callbacks.go:89]: Just NoMatch

File: pkg/operator/controller/callbacks.go (hunk 3 above)
Comment (arnongilboa) [reply pkg/operator/controller/callbacks.go]: fixed

File: pkg/operator/controller/callbacks.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (arnongilboa) [reply pkg/operator/controller/callbacks.go]: fixed

File: tests/operator_test.go [hunk 5]
```diff
@@ -438,6 +438,23 @@ var _ = Describe("ALL Operator tests", func() {
 				err = f.CdiClient.CdiV1beta1().CDIs().Delete(context.TODO(), cr.Name, metav1.DeleteOptions{DryRun: []string{"All"}})
 				Expect(err).ToNot(HaveOccurred())
 			})
+			It("[test_id:XXXX]CDI CR deletion should delete DataImportCron CRD and all DataImportCrons with finalizers", func() {
+				By("Create new DataImportCron")
+				url := fmt.Sprintf(utils.TrustedRegistryURL, f.DockerPrefix)
+				cron := NewDataImportCron("cron-test", "5Gi", scheduleEveryMinute, "ds", cdiv1.DataVolumeSourceRegistry{URL: &url, PullMethod: &registryPullNode}, cdiv1.DataImportCronRetainAll)
+				cron, err := f.CdiClient.CdiV1beta1().DataImportCrons(f.Namespace.Name).Create(context.TODO(), cron, metav1.CreateOptions{})
+				Expect(err).ToNot(HaveOccurred())
```
Comment (mhenriks) [inline tests/operator_test.go:446]: can we wait for DV to be created/succeed?  Or at least validate that the finalizer exists?

Comment (arnongilboa) [reply tests/operator_test.go:454]: fixed

**Pull request merged**